File: frontend/src/types/preferences.ts
// Email notification preferences
export interface EmailNotificationPreferences {
  platform_announcement: boolean
  account_security: boolean
  new_feature: boolean
  marketing_promotional: boolean
}

// Per-type delivery settings for each channel.
// `email` is optional because data persisted before the digest feature may omit it.
export interface NotificationTypePreference {
  in_app: boolean
  email?: string
  web_push: boolean
}

// Global channel on/off switches
export interface NotificationChannelPreferences {
  in_app: boolean
  email: boolean
  web_push: boolean
}

// Aggregated notification preferences: global channel switches + per-type delivery.
// `types` is optional because older backend data may omit it when only `channels` was stored.
export interface NotificationPreferences {
  channels: NotificationChannelPreferences
  types?: Record<string, NotificationTypePreference>
}

// All preferences categories
export interface Preferences {
  email_notification: EmailNotificationPreferences
  notifications?: NotificationPreferences
}

// API response for preferences
export interface PreferencesResponse {
  preferences: Preferences
  updated_at?: string
}

// Request to update preferences
export interface UpdatePreferencesRequest {
  email_notification?: EmailNotificationPreferences
  notifications?: NotificationPreferences
}

File: backend/internal/models/preferences.go
package models

import (
	"time"
)

// EmailNotificationPreferences contains email notification settings
type EmailNotificationPreferences struct {
	PlatformAnnouncement bool `json:"platform_announcement"`
	AccountSecurity      bool `json:"account_security"`
	NewFeature           bool `json:"new_feature"`
	MarketingPromotional bool `json:"marketing_promotional"`
}

// NotificationTypePreference controls per-type delivery for each channel.
// Email values: "instant" | "digest" | "none"
type NotificationTypePreference struct {
	InApp   bool   `json:"in_app"`
	Email   string `json:"email"`
	WebPush bool   `json:"web_push"`
}

// NotificationChannelPreferences controls global channel on/off switches
type NotificationChannelPreferences struct {
	InApp   bool `json:"in_app"`
	Email   bool `json:"email"`
	WebPush bool `json:"web_push"`
}

// NotificationPreferences aggregates channel and per-type preferences
type NotificationPreferences struct {
	Channels NotificationChannelPreferences        `json:"channels"`
	Types    map[string]NotificationTypePreference `json:"types"`
}

// Preferences contains all user preference categories
type Preferences struct {
	EmailNotification EmailNotificationPreferences `json:"email_notification"`
	Notifications     NotificationPreferences      `json:"notifications"`
}

// UserPreferences represents the user preferences entity stored in the database
type UserPreferences struct {
	ID          string      `json:"id" db:"id"`
	UserID      string      `json:"user_id" db:"user_id"`
	Preferences Preferences `json:"preferences" db:"preferences"`
	CreatedAt   time.Time   `json:"created_at" db:"created_at"`
	UpdatedAt   time.Time   `json:"updated_at" db:"updated_at"`
	Version     int64       `json:"version" db:"version"`
}

// UpdatePreferencesRequest represents the request to update user preferences
type UpdatePreferencesRequest struct {
	EmailNotification *EmailNotificationPreferences `json:"email_notification,omitempty"`
	Notifications     *NotificationPreferences      `json:"notifications,omitempty"`
}

// PreferencesResponse represents the API response for user preferences
type PreferencesResponse struct {
	Preferences Preferences `json:"preferences"`
	UpdatedAt   time.Time   `json:"updated_at"`
}

// DefaultNotificationPreferences returns sensible default notification preferences
func DefaultNotificationPreferences() NotificationPreferences {
	return NotificationPreferences{
		Channels: NotificationChannelPreferences{
			InApp:   true,
			Email:   true,
			WebPush: false,
		},
		Types: map[string]NotificationTypePreference{
			"feed.item.created": {
				InApp:   true,
				Email:   "digest",
				WebPush: true,
			},
			"feed.reply.created": {
				InApp:   true,
				Email:   "instant",
				WebPush: true,
			},
			"team.invitation": {
				InApp:   true,
				Email:   "instant",
				WebPush: false,
			},
		},
	}
}

// DefaultPreferences returns the default preferences for a new user
func DefaultPreferences() Preferences {
	return Preferences{
		EmailNotification: EmailNotificationPreferences{
			PlatformAnnouncement: true,
			AccountSecurity:      true,
			NewFeature:           true,
			MarketingPromotional: false,
		},
		Notifications: DefaultNotificationPreferences(),
	}
}
